cmd/checkmetrics/main.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Program checkmetrics compares the results from a set of Clear Containers
+metrics results, stored in CSV files, against a set of baseline metrics
+'expectations', defined in a TOML file.
+
+It returns non zero if any of the TOML metrics are not met.
+
+it prints out a tabluated report summary at the end of the run.
+*/
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+// name is the name of the program.
+const name = "checkmetrics"
+
+// usage is the usage of the program.
+const usage = name + ` checks CSV metrics results against a TOML baseline`
+
+// The TOML basefile
+var ciBasefile *baseFile
+
+// processMetrics locates the CSV file matching each entry in the TOML
+// baseline, loads and processes it, and checks if the metrics were in range.
+// Finally it generates a summary report
+func processMetrics(context *cli.Context) error {
+	var err error
+	var finalerror error // If we fail any metric, fail globally
+	var report []string  // summary report table
+	var passes int
+	var fails int
+
+	log.Debug("processMetrics")
+
+	// Process each Metrics TOML entry one at a time
+	for _, m := range ciBasefile.Metric {
+		var thisCsv csvRecord
+
+		log.Debugf("Processing %s", m.Name)
+		fullpath := path.Join(context.GlobalString("metricsdir"), m.Name)
+		fullpath = fullpath + ".csv"
+
+		log.Debugf("Fullpath %s", fullpath)
+		err = thisCsv.load(fullpath)
+		if err != nil {
+			log.Warnf("Failed to read csv [%s]", fullpath)
+			// Record that this one did not complete successfully
+			finalerror = errors.New("Fail")
+			fails += 1
+			// Not a fatal error - continue to process any remaining files
+			continue
+		}
+
+		// Now we have both the baseline and the CSV data loaded,
+		// let's go compare them
+		var mc metricsCheck
+
+		err, summary := mc.check(m, thisCsv)
+		if err != nil {
+			log.Warnf("Check for [%s] failed", m.Name)
+			log.Warnf(" with [%s]", summary)
+			finalerror = errors.New("Fail")
```

Comment:
Not sure if you mean from check() or for the whole Metric range loop we are within...
For check(), we will print the failing summary in the log, and then summary is accumulated up for the final report. A failure is non-fatal, so we continue to process the rest of the Metrics in the range.
For the range covering Metrics, finalerror is just an indicator that we have seen at least one failure (I'll go rename that as finalErroFlag or similar now to make that clearer). The actual Failures end up being reported in the report generated by the accumulated `summary` results.
Does that clarify? If not, then elaborate again. Maybe I need to add a few comments in there to clarify?